# 크롤링해온 csv파일들을 합치고 전처리하는 과정

In [1]:
import pandas as pd

### 크롤링해온 csv가 1000개 단위로 쪼개져있기 때문에 한줄로 합칠 필요가 있음

In [3]:
# 1000개에서 10000개를 한줄로 합치기 위해서 for문으로 한꺼번에 추가해주기
dfs = [] # 데이터프레임을 저장할 리스트

for i in range(1, 11):
    df = pd.read_csv(f"Data/제주도식당_크롤링_{i}000.csv")
    dfs.append(df)

result_df = pd.concat(dfs)

In [8]:
result_df.reset_index()

index foodCategory                                             images  \
0          0            x                                                  x   
1          1       카페,디저트  ['"https://search.pstatic.net/common/?autoRota...   
2          2         두부요리                                                 []   
3          3        맥주,호프                                                 []   
4          4           양식  ['"https://search.pstatic.net/common/?autoRota...   
...      ...          ...                                                ...   
9996     995            x                                                  x   
9997     996            x                                                  x   
9998     997       치킨,닭강정  ['"https://search.pstatic.net/common/?autoRota...   
9999     998            x                                                  x   
10000    999            x                                                  x   

      rating visitReviewCount blogReviewCount  \
0          x                x               x   
1       4.49               97              44   
2       4.35              429              44   
3          x               21               3   
4       4.79              299             149   
...      ...              ...             ...   
9996       x                x               x   
9997       x                x               x   
9998       x               22               2   
9999       x                x               x   
10000      x                x               x   

                                                  review  
0                                                      x  
1      ['제 입맛에는 안맞았어요. 🥲', '세화의 작은 카페 에요. 사장님 친정하시고 커...  
2      ['맛있게 잘 먹고 왔어요^', '늦은 시간 편하게 들리기 좋아요~!', '음식이 ...  
3      ['집근처라 가끔 가는곳', '맥주랑 하이볼이랑~ 새로나온 멜론샤베트! 진짜 너무맛...  
4      ['N번째 방문 중인 봉플라봉뱅! 친절하고 맛있어서 제주 오면 꼭 들리는 곳 입니다...  
...                                                  ...  
9996                                                   x  
9997                                                   x  
9998   ['맛있어요', '맛있어요 종종 배달시켜 먹어요', '', '굿', '매운맛 매워서...  
9999                                                   x  
10000                                                  x  

[10001 rows x 7 columns]

### 잘 가져왔나 직접 크롤링 기준이 되었던 데이터와 비교해보기
직접 네이버 지도가서 확인해봐야함

In [7]:
scrapyData = pd.read_csv("Data/제주도식당_전처리.csv")

In [10]:
scrapyData.head()

사업장명 업종구분대분류   업종구분소분류                        소재지전체주소  \
0      탑동에이치지   일반음식점        기타        제주특별자치도 제주시 삼도이동 1120-2   
1    에릭스에스프레소   일반음식점        기타   제주특별자치도 제주시 구좌읍 세화리 1397-6번지   
2    일품순두부한림점   일반음식점        한식   제주특별자치도 제주시 한림읍 대림리 1845-2번지   
3  김복남맥주제주아라점   일반음식점  식육(숯불구이)          제주특별자치도 제주시 아라일동 6139   
4       봉플라봉뱅   일반음식점        기타  제주특별자치도 제주시 연동 312-57번지 정인하우스   

                               도로명전체주소         lat          lng  
0      제주특별자치도 제주시 무근성안길 16, 1층 (삼도이동)  33.5151128  126.5202001  
1           제주특별자치도 제주시 구좌읍 구좌로 77, 1층  33.5214225  126.8588743  
2         제주특별자치도 제주시 한림읍 한림상로 237, 1층  33.4192601  126.2674304  
3   제주특별자치도 제주시 중앙로 581, 에이동 1층 (아라일동)  33.4714278  126.5457848  
4  제주특별자치도 제주시 문송1길 6-1, 정인하우스 1층 (연동)  33.4879976  126.4978933

In [12]:
scrapyData[9996: 10001]

사업장명 업종구분대분류 업종구분소분류                       소재지전체주소  \
9996        신마녀떡볶이제주점   일반음식점      분식  제주특별자치도 제주시 화북일동 1948-3 해성원룸   
9997         피크닉앤와인월정   일반음식점     경양식     제주특별자치도 제주시 구좌읍 월정리 482-7   
9998   화락바베큐치킨제주삼화지구점   일반음식점   호프/통닭       제주특별자치도 제주시 도련일동 1938-3   
9999               청량   일반음식점      기타    제주특별자치도 제주시 구좌읍 세화리 1521-5   
10000            오제이이   일반음식점      기타     제주특별자치도 제주시 한경면 낙천리 812-3   

                                     도로명전체주소         lat          lng  
9996   제주특별자치도 제주시 진남로6길 44, 1층 (화북일동, 해성원룸)  33.5225409  126.5654031  
9997             제주특별자치도 제주시 구좌읍 월정3길 38, 1층  33.5544690  126.7937295  
9998        제주특별자치도 제주시 건주로4길 6-8, 1층 (도련일동)  33.5167640  126.5802858  
9999              제주특별자치도 제주시 구좌읍 구좌로 31, 1층  33.5236370  126.8548548  
10000               제주특별자치도 제주시 한경면 불그못로 222  33.2960210  126.2330118

In [13]:
result_df[5000: 5006]

foodCategory                                             images rating  \
999           한식                                                 []      x   
0              x                                                  x      x   
1             한식  ['"https://search.pstatic.net/common/?autoRota...      x   
2             한식  ['"https://search.pstatic.net/common/?autoRota...      x   
3         돼지고기구이  ['"https://search.pstatic.net/common/?autoRota...      x   
4         카페,디저트  ['"https://search.pstatic.net/common/?autoRota...   4.03   

    visitReviewCount blogReviewCount  \
999              160              96   
0                  x               x   
1                 20               7   
2                 14               4   
3                 87              11   
4                 33               2   

                                                review  
999  ['잘먹어요 항상', '갈때마다 변함없이 맛있어요\n추천합니다^^', '지나가다가 ...  
0                                                    x  
1    ['오겹살 고기도 맛있었는데 밑반찬들도 정갈하게 괜찮았습니다!:)', '친절해서 좋...  
2    ['믿고 찾는 곳', '말이 필요없는\n횟집은 여기만 갑니다', '잘먹었습니다', ...  
3    ['고기양이 많아서 여럿이서 오기에도 부담없고 양념이 너무 달지만도 않고 적절하게 ...  
4    ['올레12코스 걷다가 잠시 들른고..\n창밖 풍경이 좋았어요.', '노을이 지면 ...

In [14]:
scrapyData[5000: 5006]

사업장명 업종구분대분류 업종구분소분류                        소재지전체주소  \
5000        오늘해장국   일반음식점      한식        제주특별자치도 제주시 화북일동 1994-6   
5001         에덴식당   일반음식점      기타        제주특별자치도 제주시 일도이동 54-1번지   
5002         선봉식당   일반음식점      한식  제주특별자치도 제주시 이도이동 1183-38 (1층)   
5003           누이   일반음식점   호프/통닭        제주특별자치도 제주시 삼도일동 580-38   
5004  좋아좋아숯불갈비2호점   일반음식점      한식          제주특별자치도 제주시 연동 295-17   
5005        바람과언덕   일반음식점      기타     제주특별자치도 제주시 한경면 고산리 3581번지   

                               도로명전체주소         lat          lng  
5000    제주특별자치도 제주시 진동로 5, 101호 (화북일동)  33.5194918  126.5685271  
5001       제주특별자치도 제주시 연수로3길 14 (일도이동)  33.5026758  126.5441687  
5002        제주특별자치도 제주시 동고산로 16 (이도이동)  33.4983265  126.5326888  
5003       제주특별자치도 제주시 서광로26길 2 (삼도일동)  33.4995493  126.5241306  
5004  제주특별자치도 제주시 신대로20길 28, 101호 (연동)  33.4838199  126.4934637  
5005  제주특별자치도 제주시 한경면 노을해안로 1203-4, 1층  33.3067648  126.1690797

> 확인 결과 가져온 리뷰의 행과 기준의 데이터 행이 정확히 일치함

In [16]:
# 저장하기
result_df.to_csv("Data/제주음식점리뷰_upTo_10000.csv", index=False)

### 테스트로 10000개를 붙여보기

In [18]:
newscrapyData = scrapyData[0: 10001] 

In [37]:
newscrapyData.reset_index(inplace=True)

In [38]:
result_df.reset_index(inplace=True)

In [40]:
concat_result_df = pd.concat([newscrapyData, result_df], axis=1)

In [41]:
concat_result_df.head()

index        사업장명 업종구분대분류   업종구분소분류                        소재지전체주소  \
0      0      탑동에이치지   일반음식점        기타        제주특별자치도 제주시 삼도이동 1120-2   
1      1    에릭스에스프레소   일반음식점        기타   제주특별자치도 제주시 구좌읍 세화리 1397-6번지   
2      2    일품순두부한림점   일반음식점        한식   제주특별자치도 제주시 한림읍 대림리 1845-2번지   
3      3  김복남맥주제주아라점   일반음식점  식육(숯불구이)          제주특별자치도 제주시 아라일동 6139   
4      4       봉플라봉뱅   일반음식점        기타  제주특별자치도 제주시 연동 312-57번지 정인하우스   

                               도로명전체주소         lat          lng  index  \
0      제주특별자치도 제주시 무근성안길 16, 1층 (삼도이동)  33.5151128  126.5202001      0   
1           제주특별자치도 제주시 구좌읍 구좌로 77, 1층  33.5214225  126.8588743      1   
2         제주특별자치도 제주시 한림읍 한림상로 237, 1층  33.4192601  126.2674304      2   
3   제주특별자치도 제주시 중앙로 581, 에이동 1층 (아라일동)  33.4714278  126.5457848      3   
4  제주특별자치도 제주시 문송1길 6-1, 정인하우스 1층 (연동)  33.4879976  126.4978933      4   

  foodCategory                                             images rating  \
0            x                                                  x      x   
1       카페,디저트  ['"https://search.pstatic.net/common/?autoRota...   4.49   
2         두부요리                                                 []   4.35   
3        맥주,호프                                                 []      x   
4           양식  ['"https://search.pstatic.net/common/?autoRota...   4.79   

  visitReviewCount blogReviewCount  \
0                x               x   
1               97              44   
2              429              44   
3               21               3   
4              299             149   

                                              review  
0                                                  x  
1  ['제 입맛에는 안맞았어요. 🥲', '세화의 작은 카페 에요. 사장님 친정하시고 커...  
2  ['맛있게 잘 먹고 왔어요^', '늦은 시간 편하게 들리기 좋아요~!', '음식이 ...  
3  ['집근처라 가끔 가는곳', '맥주랑 하이볼이랑~ 새로나온 멜론샤베트! 진짜 너무맛...  
4  ['N번째 방문 중인 봉플라봉뱅! 친절하고 맛있어서 제주 오면 꼭 들리는 곳 입니다...

### concat하기 위해 resetindex해줬기 때문에 index라는게 생겼기 때문에, 그걸 제외한 필요한 컬럼들만 뽑아서 저장하기

In [49]:
concat_result_df.columns

Index(['index', '사업장명', '업종구분대분류', '업종구분소분류', '소재지전체주소', '도로명전체주소', 'lat',
       'lng', 'index', 'foodCategory', 'images', 'rating', 'visitReviewCount',
       'blogReviewCount', 'review'],
      dtype='object')

In [51]:
concat_result_df = concat_result_df[["사업장명", "업종구분대분류", "업종구분소분류", "foodCategory", "소재지전체주소", "도로명전체주소", "lat", "lng", "images", "rating", "visitReviewCount", "blogReviewCount", "review"]]

In [52]:
concat_result_df.head()

사업장명 업종구분대분류   업종구분소분류 foodCategory                        소재지전체주소  \
0      탑동에이치지   일반음식점        기타            x        제주특별자치도 제주시 삼도이동 1120-2   
1    에릭스에스프레소   일반음식점        기타       카페,디저트   제주특별자치도 제주시 구좌읍 세화리 1397-6번지   
2    일품순두부한림점   일반음식점        한식         두부요리   제주특별자치도 제주시 한림읍 대림리 1845-2번지   
3  김복남맥주제주아라점   일반음식점  식육(숯불구이)        맥주,호프          제주특별자치도 제주시 아라일동 6139   
4       봉플라봉뱅   일반음식점        기타           양식  제주특별자치도 제주시 연동 312-57번지 정인하우스   

                               도로명전체주소         lat          lng  \
0      제주특별자치도 제주시 무근성안길 16, 1층 (삼도이동)  33.5151128  126.5202001   
1           제주특별자치도 제주시 구좌읍 구좌로 77, 1층  33.5214225  126.8588743   
2         제주특별자치도 제주시 한림읍 한림상로 237, 1층  33.4192601  126.2674304   
3   제주특별자치도 제주시 중앙로 581, 에이동 1층 (아라일동)  33.4714278  126.5457848   
4  제주특별자치도 제주시 문송1길 6-1, 정인하우스 1층 (연동)  33.4879976  126.4978933   

                                              images rating visitReviewCount  \
0                                                  x      x                x   
1  ['"https://search.pstatic.net/common/?autoRota...   4.49               97   
2                                                 []   4.35              429   
3                                                 []      x               21   
4  ['"https://search.pstatic.net/common/?autoRota...   4.79              299   

  blogReviewCount                                             review  
0               x                                                  x  
1              44  ['제 입맛에는 안맞았어요. 🥲', '세화의 작은 카페 에요. 사장님 친정하시고 커...  
2              44  ['맛있게 잘 먹고 왔어요^', '늦은 시간 편하게 들리기 좋아요~!', '음식이 ...  
3               3  ['집근처라 가끔 가는곳', '맥주랑 하이볼이랑~ 새로나온 멜론샤베트! 진짜 너무맛...  
4             149  ['N번째 방문 중인 봉플라봉뱅! 친절하고 맛있어서 제주 오면 꼭 들리는 곳 입니다...

In [54]:
# 앱에서 쓸 json데이터 테스트용(10000)개만 있는거 저장
concat_result_df.to_json("Data/제주도식당Data.json", orient='records', force_ascii=False)

---
# 못가져온 데이터들 뽑아보기

In [55]:
concat_result_df[concat_result_df['review'] == "x"]

사업장명 업종구분대분류 업종구분소분류 foodCategory  \
0            탑동에이치지   일반음식점      기타            x   
6             가빈거래처   일반음식점      기타            x   
13            길섶나그네   일반음식점      한식            x   
14              사랑뜰   일반음식점      기타            x   
15             불티나네   일반음식점      기타            x   
...             ...     ...     ...          ...   
9994   부산아지매국밥제주시청점   일반음식점      한식            x   
9996      신마녀떡볶이제주점   일반음식점      분식            x   
9997       피크닉앤와인월정   일반음식점     경양식            x   
9999             청량   일반음식점      기타            x   
10000          오제이이   일반음식점      기타            x   

                             소재지전체주소                                도로명전체주소  \
0            제주특별자치도 제주시 삼도이동 1120-2        제주특별자치도 제주시 무근성안길 16, 1층 (삼도이동)   
6            제주특별자치도 제주시 아라일동 368-81       제주특별자치도 제주시 산천단동2길 13, 1층 (아라일동)   
13        제주특별자치도 제주시 조천읍 와산리 1303-6            제주특별자치도 제주시 조천읍 미래로 241, 1층   
14        제주특별자치도 제주시 한림읍 한림리 1419-1          제주특별자치도 제주시 한림읍 한림로 662-1, 1층   
15         제주특별자치도 제주시 애월읍 하귀2리 2720        제주특별자치도 제주시 애월읍 가문동상2길 29-7, 1층   
...                              ...                                    ...   
9994   제주특별자치도 제주시 이도이동 1055-11 석현원룸     제주특별자치도 제주시 오복1길 9, 석현원룸 1층 (이도이동)   
9996    제주특별자치도 제주시 화북일동 1948-3 해성원룸  제주특별자치도 제주시 진남로6길 44, 1층 (화북일동, 해성원룸)   
9997       제주특별자치도 제주시 구좌읍 월정리 482-7            제주특별자치도 제주시 구좌읍 월정3길 38, 1층   
9999      제주특별자치도 제주시 구좌읍 세화리 1521-5             제주특별자치도 제주시 구좌읍 구좌로 31, 1층   
10000      제주특별자치도 제주시 한경면 낙천리 812-3               제주특별자치도 제주시 한경면 불그못로 222   

              lat          lng images rating visitReviewCount blogReviewCount  \
0      33.5151128  126.5202001      x      x                x               x   
6      33.4491844  126.5581602      x      x                x               x   
13     33.4683304  126.6786072      x      x                x               x   
14     33.4143390  126.2639232      x      x                x               x   
15     33.4845981  126.3909537      x      x                x               x   
...           ...          ...    ...    ...              ...             ...   
9994   33.4961120  126.5340393      x      x                x               x   
9996   33.5225409  126.5654031      x      x                x               x   
9997   33.5544690  126.7937295      x      x                x               x   
9999   33.5236370  126.8548548      x      x                x               x   
10000  33.2960210  126.2330118      x      x                x               x   

      review  
0          x  
6          x  
13         x  
14         x  
15         x  
...      ...  
9994       x  
9996       x  
9997       x  
9999       x  
10000      x  

[4209 rows x 13 columns]

> 이런식으로 했을 때 4209개를 못가져온 것으로 확인됨   
추후에 이 데이터에대해 보완하던가 말던가 필요

---
### 10000개 이후의 데이터 합치기
- 간편하게 하기 위해서 제주도식당_크롤링_20656.csv -> 제주도식당_크롤링_20000.csv 변경

In [4]:
# 1000개에서 10000개를 한줄로 합치기 위해서 for문으로 한꺼번에 추가해주기
dfs = [] # 데이터프레임을 저장할 리스트

for i in range(1, 21):
    df = pd.read_csv(f"Data/제주도식당_크롤링_{i}000.csv")
    dfs.append(df)

result_df = pd.concat(dfs)

In [5]:
result_df.count(axis=1)

0       6
1       6
2       6
3       6
4       6
       ..
1650    6
1651    6
1652    6
1653    6
1654    6
Length: 20656, dtype: int64

### index 안꼬이고 잘 가져왔나 확인하기

In [9]:
result_df.reset_index(inplace=True)

In [10]:
scrapyData.reset_index(inplace=True)

In [15]:
result_df[20650: 20657]

index foodCategory                                             images  \
20650   1649       카페,디저트  ['"https://search.pstatic.net/common/?autoRota...   
20651   1650            x                                                  x   
20652   1651       카페,디저트  ['"https://search.pstatic.net/common/?autoRota...   
20653   1652          돈가스  ['"https://search.pstatic.net/common/?autoRota...   
20654   1653       카페,디저트  ['"https://search.pstatic.net/common/?autoRota...   
20655   1654            x                                                  x   

      rating visitReviewCount blogReviewCount  \
20650      x             1136             942   
20651      x                x               x   
20652      x               37               6   
20653   4.58             6598            4155   
20654      x               97             157   
20655      x                x               x   

                                                  review  
20650  ['', '커피가 맛있어요^^제주도 여행와서 힐링하고가요^^', '점심 먹고 방문한...  
20651                                                  x  
20652  ['좋아요', '맛있어요 ^^', '통귤이 생각보다 작았지만 맛있었어요!!!', '...  
20653  ['돈까스를 떠나서 말라비틀어진 단무지와 밥 진짜 맛없는 깍두기 너무나도 평범한 샐...  
20654  ['표선면 가시리에 있는 수목원카페입니다.\n입장료가 있지만 도민에게 40%할인되고...  
20655                                                  x

In [16]:
scrapyData[20650: 20657]

index      사업장명 업종구분대분류 업종구분소분류                       소재지전체주소  \
20650  20650   노바운더리카페   휴게음식점     커피숍   제주특별자치도 서귀포시 표선면 성읍리 2375-1   
20651  20651   못(MOAT)   휴게음식점     커피숍  제주특별자치도 서귀포시 남원읍 남원리 2406-10   
20652  20652  제주하이루탕후루   휴게음식점  일반조리판매       제주특별자치도 서귀포시 서귀동 271-40   
20653  20653        연돈   휴게음식점  일반조리판매       제주특별자치도 서귀포시 색달동 2132-2   
20654  20654    가시림 카페   휴게음식점     커피숍     제주특별자치도 서귀포시 표선면 가시리 3275   
20655  20655      올드타운   휴게음식점      다방    제주특별자치도 서귀포시 표선면 성읍리 987-1   

                                     도로명전체주소         lat          lng  
20650        제주특별자치도 서귀포시 표선면 번영로 2610, 1.2층  33.4018163  126.7831445  
20651  제주특별자치도 서귀포시 남원읍 태위로510번길 47-1, 1, 2층  33.2724596  126.6984028  
20652      제주특별자치도 서귀포시 중정로73번길 20, 1층 (서귀동)  33.2498829  126.5639792  
20653          제주특별자치도 서귀포시 색달로 10, 2층 (색달동)  33.2588634  126.4061156  
20654            제주특별자치도 서귀포시 표선면 녹산로5번길 171  33.3610266  126.7516155  
20655  제주특별자치도 서귀포시 표선면 성읍정의현로22번길 18-8, 20호  33.3844221  126.8031138

> 17000 ~ 17010은 꼬여있음

> 끝에는 동일함

In [68]:
result_df.reset_index(inplace=True)
print(result_df[19000: 19010])
print(result_df[20000: 20010])
# result_df[19000: 19010]

       index foodCategory                                             images  \
19000    999           카페  ['"https://search.pstatic.net/common/?autoRota...   
19001   1000            x                                                  x   
19002   1001       카페,디저트  ['"https://search.pstatic.net/common/?autoRota...   
19003   1002           카페  ['"https://search.pstatic.net/common/?autoRota...   
19004   1003            x                                                  x   
19005   1004            x                                                  x   
19006   1005            x                                                  x   
19007   1006            x                                                  x   
19008   1007            x                                                  x   
19009   1008       카페,디저트  ['"https://search.pstatic.net/common/?autoRota...   

      rating visitReviewCount blogReviewCount  \
19000      x              126               0   
19001      x         

In [71]:
result_df.count(axis=1)

0        7
1        7
2        7
3        7
4        7
        ..
19651    7
19652    7
19653    7
19654    7
19655    7
Length: 19656, dtype: int64

---
# 식당 크롤링 데이터 전처리하기
- 15001 ~ 19000 까지의 데이터는 제외

In [3]:
# 기준 데이터 불러오기
scrapyData = pd.read_csv("Data/제주도식당_전처리.csv")

In [10]:
# 15001 ~ 19000 제외하기
# 슬라이싱을 사용하여 특정 범위의 행을 제외하고 다른 행들만 선택
resultData = scrapyData.drop(scrapyData.index[15000:19000])

In [11]:
resultData.count(axis=1)

0        7
1        7
2        7
3        7
4        7
        ..
20651    7
20652    7
20653    7
20654    7
20655    7
Length: 16656, dtype: int64

In [13]:
# 1000개에서 15000개를 한줄로 합치기 위해서 for문으로 한꺼번에 추가해주기
dfs = [] # 데이터프레임을 저장할 리스트

for i in range(1, 16):
    df = pd.read_csv(f"Data/크롤링 데이터/제주도식당_크롤링_{i}000.csv")
    dfs.append(df)

reviewData = pd.concat(dfs)

In [16]:
reviewData.reset_index(inplace=True)

In [17]:
reviewData.tail()

index foodCategory                                             images  \
14996    995            x                                                  x   
14997    996       카페,디저트  ['"https://search.pstatic.net/common/?autoRota...   
14998    997       스테이크,립  ['"https://search.pstatic.net/common/?autoRota...   
14999    998            x                                                  x   
15000    999          일식당  ['"https://search.pstatic.net/common/?autoRota...   

      rating visitReviewCount blogReviewCount  \
14996      x                x               x   
14997      x               31               5   
14998      x               65               6   
14999      x                x               x   
15000      x               69              14   

                                                  review  
14996                                                  x  
14997  ['볼일이 있어 이곳 화순에 처음 왔는데, 가까이 이프유캔 카페있어 들렀어요~\n바...  
14998  ['분위기 좋고 무엇보다 스테이크 넘 맛나요 사장님이 꼼꼼하게 서비스해주셔서 편하고...  
14999                                                  x  
15000  ['양도 많고 너무 맛있어요', '시키기 전엔 좀 비싸네? 싶었는데 음식 나오니 이...

In [20]:
resultData[14996: 15001]

사업장명 업종구분대분류          업종구분소분류                     소재지전체주소  \
14996  중문클럽하우스주식회사   일반음식점               한식       제주특별자치도 서귀포시 색달동 2101   
14997         이프유캔   일반음식점              경양식  제주특별자치도 서귀포시 안덕면 화순리 375-1   
14998   더허그스테이크하우스   일반음식점  외국음식전문점(인도,태국등)     제주특별자치도 서귀포시 서홍동 253-23   
14999   폰드메르라운지&델리   일반음식점               한식     제주특별자치도 서귀포시 색달동 3039-1   
19000        호로록몽땅   휴게음식점            패스트푸드     제주특별자치도 제주시 노형동 2535-10   

                                    도로명전체주소         lat          lng  
14996       제주특별자치도 서귀포시 중문관광로72번길 60 (색달동)  33.2521851  126.4055494  
14997            제주특별자치도 서귀포시 안덕면 화순중앙로 109  33.2423372  126.3357040  
14998            제주특별자치도 서귀포시 중앙로 207 (서홍동)  33.2631201  126.5585981  
14999  제주특별자치도 서귀포시 중문관광로72번길 100, 1층 (색달동)  33.2441577  126.4056010  
19000         제주특별자치도 제주시 노형로 363, 2층 (노형동)  33.4830850  126.4748391

> 15000까지는 전부 일치함

In [21]:
# 19000 ~ 20565 붙여주기
df20 = pd.read_csv(f"Data/크롤링 데이터/제주도식당_크롤링_20000.csv")

In [23]:
df20.tail()

foodCategory                                             images rating  \
1650            x                                                  x      x   
1651       카페,디저트  ['"https://search.pstatic.net/common/?autoRota...      x   
1652          돈가스  ['"https://search.pstatic.net/common/?autoRota...   4.58   
1653       카페,디저트  ['"https://search.pstatic.net/common/?autoRota...      x   
1654            x                                                  x      x   

     visitReviewCount blogReviewCount  \
1650                x               x   
1651               37               6   
1652             6598            4155   
1653               97             157   
1654                x               x   

                                                 review  
1650                                                  x  
1651  ['좋아요', '맛있어요 ^^', '통귤이 생각보다 작았지만 맛있었어요!!!', '...  
1652  ['돈까스를 떠나서 말라비틀어진 단무지와 밥 진짜 맛없는 깍두기 너무나도 평범한 샐...  
1653  ['표선면 가시리에 있는 수목원카페입니다.\n입장료가 있지만 도민에게 40%할인되고...  
1654                                                  x

In [26]:
finalResultData = pd.concat([reviewData, df20], axis=0)

In [29]:
# 마지막 체크하기
finalResultData.tail()
finalResultData[15001: 15010]

index foodCategory                                             images  \
0    NaN       카페,디저트  ['"https://search.pstatic.net/common/?autoRota...   
1    NaN       카페,디저트  ['"https://search.pstatic.net/common/?autoRota...   
2    NaN       카페,디저트  ['"https://search.pstatic.net/common/?autoRota...   
3    NaN       카페,디저트  ['"https://search.pstatic.net/common/?autoRota...   
4    NaN       카페,디저트  ['"https://search.pstatic.net/common/?autoRota...   
5    NaN            x                                                  x   
6    NaN            x                                                  x   
7    NaN            x                                                  x   
8    NaN            x                                                  x   

  rating visitReviewCount blogReviewCount  \
0      x             2306            3637   
1      x               32               9   
2      x              204             165   
3   4.39              350             169   
4      x              138              13   
5      x                x               x   
6      x                x               x   
7      x                x               x   
8      x                x               x   

                                              review  
0  ['제주도 여행와서 너무 멋진 카페 발견했어요^^\n뷰도 너무 예쁘고 해안가 산책도...  
1  ['연삼로 큰길에 이렇게 분위기 좋은 카페가 있었다니!!\n구옥을 예쁘게 리모델링해...  
2  ['치즈케이크 맛집으로 유명한 청연디저트.ᐟ\n치즈케이크도 종류가 다양한지 여기서 ...  
3  ['저 혼자 간거라 스콘 몇 개 구매한 게 다지만 정말 맛있네요! 계란 비린맛도 안...  
4  ['티그레맛집이에요!', '가까워서 종종 들렀어요! 여기 디저트 찐맛집입니다!! 🥺...  
5                                                  x  
6                                                  x  
7                                                  x  
8                                                  x

In [30]:
resultData.tail()
resultData[15001: 15010]

사업장명 업종구분대분류 업종구분소분류                    소재지전체주소  \
19001             웨이뷰   휴게음식점     커피숍  제주특별자치도 제주시 한림읍 옹포리 419-3   
19002             담담히   휴게음식점     커피숍      제주특별자치도 제주시 오라이동 1224   
19003              청연   휴게음식점     커피숍    제주특별자치도 제주시 아라일동 6034-5   
19004             구움몽   휴게음식점     커피숍   제주특별자치도 제주시 한림읍 협재리 1732   
19005              새록   휴게음식점     커피숍    제주특별자치도 제주시 아라일동 6126-2   
19006  빵장수단팥빵카페기소담제주점   휴게음식점   패스트푸드    제주특별자치도 제주시 노형동 1062-10   
19007    세븐일레븐제주삼화센터점   휴게음식점     편의점    제주특별자치도 제주시 도련일동 1940-9   
19008       씨유제주서부중앙점   휴게음식점     편의점       제주특별자치도 제주시 해안동 2374   
19009            도넛킹2   휴게음식점  일반조리판매      제주특별자치도 제주시 일도일동 1137   

                                   도로명전체주소         lat          lng  
19001      제주특별자치도 제주시 한림읍 옹포7길 25-3, 1~2층  33.4072901  126.2526795  
19002       제주특별자치도 제주시 연삼로 200, 1층 (오라이동)  33.4924009  126.5136398  
19003  제주특별자치도 제주시 도남로 189, 1층 104호 (아라일동)  33.4857160  126.5344816  
19004          제주특별자치도 제주시 한림읍 한림로 359, 1층  33.3952796  126.2419565  
19005      제주특별자치도 제주시 인다11길 28, 1층 (아라일동)  33.4738263  126.5504918  
19006         제주특별자치도 제주시 수덕5길 6, 1층 (노형동)  33.4870626  126.4770456  
19007    제주특별자치도 제주시 건주로4길 10-3, 1층 (도련일동)  33.5163366  126.5795199  
19008          제주특별자치도 제주시 노형로 2, 1층 (해안동)  33.4606608  126.4482981  
19009     제주특별자치도 제주시 동문로4길 1, 1~2층 (일도일동)  33.5126903  126.5287476

--- 
### finalResultData와 resultData 합치기
- 리뷰 못가져온 장소는 지워버리기

In [32]:
finalResultData.reset_index(inplace=True)

In [33]:
resultData.reset_index(inplace=True)

In [35]:
concaatResultData = pd.concat([resultData, finalResultData], axis=1)

In [38]:
concaatResultData.tail()

index      사업장명 업종구분대분류 업종구분소분류                       소재지전체주소  \
16651  20651   못(MOAT)   휴게음식점     커피숍  제주특별자치도 서귀포시 남원읍 남원리 2406-10   
16652  20652  제주하이루탕후루   휴게음식점  일반조리판매       제주특별자치도 서귀포시 서귀동 271-40   
16653  20653        연돈   휴게음식점  일반조리판매       제주특별자치도 서귀포시 색달동 2132-2   
16654  20654    가시림 카페   휴게음식점     커피숍     제주특별자치도 서귀포시 표선면 가시리 3275   
16655  20655      올드타운   휴게음식점      다방    제주특별자치도 서귀포시 표선면 성읍리 987-1   

                                     도로명전체주소         lat          lng  \
16651  제주특별자치도 서귀포시 남원읍 태위로510번길 47-1, 1, 2층  33.2724596  126.6984028   
16652      제주특별자치도 서귀포시 중정로73번길 20, 1층 (서귀동)  33.2498829  126.5639792   
16653          제주특별자치도 서귀포시 색달로 10, 2층 (색달동)  33.2588634  126.4061156   
16654            제주특별자치도 서귀포시 표선면 녹산로5번길 171  33.3610266  126.7516155   
16655  제주특별자치도 서귀포시 표선면 성읍정의현로22번길 18-8, 20호  33.3844221  126.8031138   

       level_0  index foodCategory  \
16651     1650    NaN            x   
16652     1651    NaN       카페,디저트   
16653     1652    NaN          돈가스   
16654     1653    NaN       카페,디저트   
16655     1654    NaN            x   

                                                  images rating  \
16651                                                  x      x   
16652  ['"https://search.pstatic.net/common/?autoRota...      x   
16653  ['"https://search.pstatic.net/common/?autoRota...   4.58   
16654  ['"https://search.pstatic.net/common/?autoRota...      x   
16655                                                  x      x   

      visitReviewCount blogReviewCount  \
16651                x               x   
16652               37               6   
16653             6598            4155   
16654               97             157   
16655                x               x   

                                                  review  
16651                                                  x  
16652  ['좋아요', '맛있어요 ^^', '통귤이 생각보다 작았지만 맛있었어요!!!', '...  
16653  ['돈까스를 떠나서 말라비틀어진 단무지와 밥 진짜 맛없는 깍두기 너무나도 평범한 샐...  
16654  ['표선면 가시리에 있는 수목원카페입니다.\n입장료가 있지만 도민에게 40%할인되고...  
16655                                                  x

In [39]:
# 필요 컬럼만 남기기
concaatResultData = concaatResultData[["사업장명", "업종구분대분류", "업종구분소분류", "foodCategory", "소재지전체주소", "도로명전체주소", "lat", "lng", "images", "rating", "visitReviewCount", "blogReviewCount", "review"]]

In [73]:
# 결측치 확인
concaatResultData.isna().count()

사업장명                16656
업종구분대분류             16656
업종구분소분류             16656
foodCategory        16656
소재지전체주소             16656
도로명전체주소             16656
lat                 16656
lng                 16656
images              16656
rating              16656
visitReviewCount    16656
blogReviewCount     16656
review              16656
dtype: int64

In [74]:
# null값 확인
concaatResultData.isnull().sum()

사업장명                 0
업종구분대분류              0
업종구분소분류              3
foodCategory         1
소재지전체주소             68
도로명전체주소              0
lat                  0
lng                  0
images               0
rating               0
visitReviewCount     0
blogReviewCount      0
review               0
dtype: int64

In [75]:
concaatResultData = concaatResultData.fillna("x")

In [76]:
# null값 확인
concaatResultData.isnull().sum()

사업장명                0
업종구분대분류             0
업종구분소분류             0
foodCategory        0
소재지전체주소             0
도로명전체주소             0
lat                 0
lng                 0
images              0
rating              0
visitReviewCount    0
blogReviewCount     0
review              0
dtype: int64

In [48]:
# 전부 못가져온거 확인하기
concaatResultData[concaatResultData['review'] == "x"].count(axis=1)

0        13
6        13
13       13
14       13
15       13
         ..
16640    13
16642    13
16645    13
16651    13
16655    13
Length: 7798, dtype: int64

> 16656개의 데이터중 8858개의 리뷰만 존재, 7798개는 못가져옴

### 리뷰 na값인 경우에도 지도데이터로는 사용이 가능하니 넣어주고   
### na값을 지운건 학습용 데이터로 사용하기

In [77]:
# Map에 쓸 데이터는 리뷰가 필요 없기 때문에 제거
mapData = concaatResultData.drop(columns=['review'])

In [78]:
mapData[mapData['foodCategory'] == "x"].count(axis=1)

0        12
6        12
13       12
14       12
15       12
         ..
16640    12
16642    12
16645    12
16651    12
16655    12
Length: 7799, dtype: int64

In [79]:
mapData[13876: 13877]

사업장명 업종구분대분류 업종구분소분류 foodCategory                소재지전체주소  \
13876  뽕이네각재기 3호점서귀포   일반음식점      한식            x  제주특별자치도 서귀포시 서홍동 1348   

                                도로명전체주소         lat          lng images  \
13876  제주특별자치도 서귀포시 일주동로 8806, 1층 (서홍동)  33.2541295  126.5482179     []   

      rating visitReviewCount blogReviewCount  
13876   4.62              140              46

In [80]:
# 지도에 들어갈 모든 음식점 데이터 저장
mapData.to_json("Data/제주도식당_지도Data.json", orient='records', force_ascii=False)

In [81]:
# 학습용에 쓸 데이터는 리뷰가 x인건 제외하기
mLData = concaatResultData[concaatResultData['review'] != "x"]

In [62]:
mLData.tail()

사업장명 업종구분대분류 업종구분소분류 foodCategory                      소재지전체주소  \
16649  배스킨라빈스서귀포표선점   휴게음식점   아이스크림        아이스크림   제주특별자치도 서귀포시 표선면 표선리 633-2   
16650       노바운더리카페   휴게음식점     커피숍       카페,디저트  제주특별자치도 서귀포시 표선면 성읍리 2375-1   
16652      제주하이루탕후루   휴게음식점  일반조리판매       카페,디저트      제주특별자치도 서귀포시 서귀동 271-40   
16653            연돈   휴게음식점  일반조리판매          돈가스      제주특별자치도 서귀포시 색달동 2132-2   
16654        가시림 카페   휴게음식점     커피숍       카페,디저트    제주특별자치도 서귀포시 표선면 가시리 3275   

                                 도로명전체주소         lat          lng  \
16649         제주특별자치도 서귀포시 표선면 표선동서로 236  33.3248679  126.8324855   
16650    제주특별자치도 서귀포시 표선면 번영로 2610, 1.2층  33.4018163  126.7831445   
16652  제주특별자치도 서귀포시 중정로73번길 20, 1층 (서귀동)  33.2498829  126.5639792   
16653      제주특별자치도 서귀포시 색달로 10, 2층 (색달동)  33.2588634  126.4061156   
16654        제주특별자치도 서귀포시 표선면 녹산로5번길 171  33.3610266  126.7516155   

                                                  images rating  \
16649  ['"https://search.pstatic.net/common/?autoRota...      x   
16650  ['"https://search.pstatic.net/common/?autoRota...      x   
16652  ['"https://search.pstatic.net/common/?autoRota...      x   
16653  ['"https://search.pstatic.net/common/?autoRota...   4.58   
16654  ['"https://search.pstatic.net/common/?autoRota...      x   

      visitReviewCount blogReviewCount  \
16649              365              13   
16650             1136             942   
16652               37               6   
16653             6598            4155   
16654               97             157   

                                                  review  
16649  ['굿', '좋아요', '굳', '역시 베스', '굿', '굿', '굿', '굿긋'...  
16650  ['', '커피가 맛있어요^^제주도 여행와서 힐링하고가요^^', '점심 먹고 방문한...  
16652  ['좋아요', '맛있어요 ^^', '통귤이 생각보다 작았지만 맛있었어요!!!', '...  
16653  ['돈까스를 떠나서 말라비틀어진 단무지와 밥 진짜 맛없는 깍두기 너무나도 평범한 샐...  
16654  ['표선면 가시리에 있는 수목원카페입니다.\n입장료가 있지만 도민에게 40%할인되고...

In [82]:
# 학습용 음식점 데이터 저장
mLData.to_csv("Data/제주도식당_학습용Data.csv", index=False)